pkg/intercp/catalog/writer.go

```diff
@@ -0,0 +1,67 @@
+package catalog
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/pkg/core/user"
+)
+
+var writerLog = core.Log.WithName("intercp").WithName("catalog").WithName("writer")
+
+type catalogWriter struct {
+	catalog    Catalog
+	heartbeats *Heartbeats
+	instance   Instance
+	interval   time.Duration
+}
+
+var _ component.Component = &catalogWriter{}
+
+func NewWriter(catalog Catalog, heartbeats *Heartbeats, instance Instance, interval time.Duration) component.Component {
+	leaderInstance := instance
+	leaderInstance.Leader = true
+	return &catalogWriter{
+		catalog:    catalog,
+		heartbeats: heartbeats,
+		instance:   leaderInstance,
+		interval:   interval,
+	}
+}
+
+func (r *catalogWriter) Start(stop <-chan struct{}) error {
+	heartbeatLog.Info("starting catalog writer")
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+	writerLog.Info("replacing a leader in the catalog")
+	if err := r.catalog.ReplaceLeader(ctx, r.instance); err != nil {
+		writerLog.Error(err, "could not replace leader") // continue, it will be replaced in ticker anyways
+	}
+	ticker := time.NewTicker(r.interval)
+	for {
+		select {
+		case <-ticker.C:
+			instances := r.heartbeats.ResetAndCollect()
+			instances = append(instances, r.instance)
+			sort.Stable(InstancesByID(instances))
```

Comment:
I'd move `sort.Stable` inside `catalog.Replace` so only `ConfigCatalog` cares about ordering, not its clients.

Reply:
good point. Thanks, moved.